credentials/alts/alts_test.go

```diff
@@ -300,3 +353,49 @@ func versions(minMajor, minMinor, maxMajor, maxMinor uint32) *altspb.RpcProtocol
 		MaxRpcVersion: version(maxMajor, maxMinor),
 	}
 }
+
+func startFakeHandshakerService(t *testing.T, wait *sync.WaitGroup) (stop func(), address string) {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatalf("Failed to listen on address %s: %v", address, err)
+	}
+	s := grpc.NewServer()
+	altspb.RegisterHandshakerServiceServer(s, &testutil.FakeHandshaker{})
+	wait.Add(1)
+	go func() {
+		defer wait.Done()
+		if err := s.Serve(listener); err != nil {
+			t.Errorf("failed to serve: %v", err)
+		}
+	}()
+	return func() { s.Stop() }, listener.Addr().String()
+}
+
+func startServer(t *testing.T, handshakerServiceAddress string, wait *sync.WaitGroup) (stop func(), address string) {
+	lis, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Errorf("net.Listen(tcp, :0) failed: %v", err)
+	}
+	serverOpts := &ServerOptions{HandshakerServiceAddress: handshakerServiceAddress}
+	creds := NewServerCreds(serverOpts)
+	s := grpc.NewServer(grpc.Creds(creds))
+	testpb.RegisterTestServiceServer(s, &testServer{})
+	wait.Add(1)
+	go func() {
+		defer wait.Done()
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("s.Serve(%v) failed: %v", lis, err)
+		}
+	}()
+	return func() { s.Stop() }, lis.Addr().String()
+}
+
+type testServer struct {
+	testpb.UnimplementedTestServiceServer
+}
+
+func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
```

Comment:
You could use our stubserver type here: https://github.com/grpc/grpc-go/blob/3489bb7d51741618f1b1e0c723ca6dc905118e3b/internal/stubserver/stubserver.go#L42, and thereby get rid of `startServer`. Would that be possible?

Reply:
Thanks for the pointer! I kept the `startServer` function (just to consolidate the logic of starting up the server), but used `StubServer` to implement it, which greatly simplifies the logic. :)

Reply:
Sorry for all the commits: I tried doing this but it ended being super flaky, so reverted back to the original solution.